addons/web/static/src/js/views/file_upload_mixin.js

```diff
@@ -0,0 +1,217 @@
+            if (predicate && !predicate(upload)) {
+                continue;
+            }
+            let $targetCard;
+            if (upload.progressCard) {
+                await upload.progressCard.prependTo(this.renderer.$el);
+                $targetCard = upload.progressCard.$el;
+            } else if (targetCallback) {
+                $targetCard = targetCallback(upload);
+            }
+            await upload.progressBar.appendTo($targetCard);
+        }
+    },
+    /**
+     * used to use a mocked version of Xhr in the tests.
+     *
+     * @private
+     * @returns {XMLHttpRequest}
+     */
+    _createXhr() {
+        return new window.XMLHttpRequest();
+    },
+    /**
+     * Creates a progress bar and a progress card and add them to the uploads.
+     *
+     * @private
+     * @param {integer} fileUploadId
+     * @param {Object} params
+     * @param {integer} params.recordId
+     * @param {string} params.title
+     * @param {string} params.type file mimetype
+     * @param {Object} params.xhr XMLHttpRequest
+     */
+    _makeFileUploadProgressBar(fileUploadId, params) {
+        let progressCard;
+        if (!params.recordId || this.viewType !== 'kanban') {
+            progressCard = new ProgressCard(this, {
+                title: params.title,
+                type: params.type,
+                viewType: this.viewType,
+            });
+        }
+        const progressBar = new ProgressBar(this, {
+            xhr: params.xhr,
+            title: params.title,
+            fileUploadId,
+        });
+        // the event listener is added here outside of the widgets as adding them outside of the scope of the definition
+        // prevents the event to be properly listened to (the issue seems to be related to the behaviour of xhr).
+        params.xhr.upload.addEventListener("progress", ev => {
+            if (ev.lengthComputable) {
+                progressCard && progressCard.update(ev.loaded, ev.total);
+                progressBar.update(ev.loaded, ev.total);
+            }
+        });
+
+        const data = Object.assign({ progressBar, progressCard }, params);
+        delete data.title;
+        delete data.type;
+        delete data.xhr;
+        this._addUpload(fileUploadId, data);
+    },
+    /**
+     * Prepares files for upload.
+     *
+     * @private
+     * @param {Object[]} files
+     */
+     async _processFiles(files) {
+        const fileUploadId = _.uniqueId('fileUploadId');
+        let title = _.str.sprintf(_t('%s Files'), files.length);
+        let type;
+        if (files.length === 1) {
+            title = files[0].name;
+            type = files[0].type;
+        }
+
+        await this._uploadFiles('/web/binary/upload_attachment', files, {
+            onUpload: async xhr => {
+                this._makeFileUploadProgressBar(fileUploadId, {xhr, title, type});
+                await this._attachFileUploadProgressBars();
```

Comment:
This seems to be called every time a new progress bar is made, so maybe this should be included at the end of implementation of `_makeFileUploadProgressBar`. Note that the special stuff in  `document_controller_mixin` that uses `_attachProgressBars` instead should no longer be necessary if `_processFiles` (to be called `_uploadFiles` IMO) uses overridable upload options that allow this behaviour from `_attachFileUploadProgressBars()`.

Reply:
I made a separate commit for these changes